### PR TITLE
Cache extracted codings in bundle retrieves

### LINKED ISSUE #1450: BundleDataSource: cache extracted codings per resource and avoid reflection/re-extraction on every retrieve
## Problem

`BundleDataSource.Retrieve<T>` with a code or value-set filter re-does per-candidate work on **every** retrieve over the same bundle (`Cql/Cql.Firely/BundleDataSource.cs`, `Cql/Cql.Firely/IndexedBundle.cs`):

- `IndexedBundle.FilterByType<T>(filter)` calls `codedCandidate.ToCodings()` per candidate per retrieve, re-allocating the coding enumeration each time.
- When a `CodeProperty` is specified, `ExecuteFilter` reads it via `PropertyInfo.GetValue(instance)` (reflection) per candidate per retrieve, then converts with `ToCodings()`.
- `RetrieveByCodes` with an explicit code list checks each candidate coding with `allowedCodes.Any(...)` — `O(codes × codings)` per candidate.

A measure evaluation (e.g. HEDIS) issues many retrieves over the same immutable bundle — commonly dozens of value-set-filtered retrieves against the same resource types — so the same codings are extracted, allocated, and compared over and over.

## Proposal

The bundle inside `IndexedBundle` is immutable for the lifetime of the data source, so extraction work can be done once:

1. **Cache extracted codings per resource.** In `IndexedBundle`, memoize the coding list per candidate per code source — keyed by type for the `ICoded` path and by (type, code property) for the `CodeProperty` path — so each resource's codings are extracted at most once per bundle lifetime, regardless of how many retrieves hit that type.
2. **Compile code-property accessors.** Replace `PropertyInfo.GetValue` with cached compiled delegates (expression trees, cached statically per `PropertyInfo`), eliminating per-candidate reflection.
3. **Hash-set lookup for explicit code lists.** In `RetrieveByCodes`, when the default string comparers are in use, build a set keyed by (system, code) once per retrieve instead of the nested `Any(...)` scan. Custom comparers keep the existing fallback path so semantics are preserved exactly, including null-system handling.

Behavioral requirements:

- Bit-for-bit parity with current results, including: case-insensitive default comparison, null system/code handling, at-most-once yielding per resource, derived-type indexing, and the `ICoded` vs `CodeProperty` paths (both scalar `DataType` and `IEnumerable<DataType>` properties).
- Thread-safe lazy initialization of the caches, so a data source can be shared by concurrent evaluations.

## Context

Identified while investigating Firely Server measure-evaluation performance (see also #1440 for the assembly-load/JIT side). For population-level HEDIS reports the retrieve path is one of the dominant per-subject runtime costs after JIT.

### PR DESCRIPTION
**Pull request by alexzautke**
## Description

Retrieves with a code or value-set filter re-did per-candidate work on every retrieve over the same immutable bundle: `IndexedBundle` re-extracted each candidate's codings (allocating on every call), code properties were read via `PropertyInfo.GetValue` reflection per candidate, and explicit code lists were matched by scanning the list for every coding (`O(codes × codings)`). A measure evaluation issues many retrieves over the same bundle, so the same extraction and comparison work ran over and over.

This PR makes that work run once per bundle:

- **`IndexedBundle`** memoizes each resource's extracted codings per retrieved type (`ICoded` path) and per (type, code property) — built lazily and thread-safely on the first coded retrieve of that key, reused by all later retrieves. Filtering runs over the cached `Coding[]` arrays.
- **`CompiledPropertyAccessor`** (new, internal) replaces `PropertyInfo.GetValue` with compiled expression-tree getters, cached statically and keyed by the resolved getter `MethodInfo` — so the `FhirModelPropertyInfo` wrappers that are allocated fresh on every lookup share one compiled getter instead of growing the cache unboundedly. The getter guards with a type test so instances that lack the property yield no codings (matching `FhirModelPropertyInfo.GetValue`'s behavior) rather than throwing.
- **`RetrieveByCodes`** matches explicit code lists via a `(system, code)` hash set when the default string comparers are in use. The set reproduces the default `Equivalent` semantics exactly: NFC-normalized, case-insensitive comparison where null only matches null. Custom comparers keep the original scan path unchanged.

No public API changes — everything added is `internal`. Behavior parity is preserved, including at-most-once yielding per resource, base-type retrieves, and null system/code handling.

## Related issues

Closes #1450.

## Testing

- 25 new tests in `CoreTests/Fhir/BundleDataSourceRetrieveTests.cs`: parity for the `ICoded` and code-property paths (scalar `DataType` and `IEnumerable<DataType>`), explicit code lists with null systems and case differences, at-most-once yield, empty bundle, base-type retrieves, cache-key discrimination per type/property, `IValueSetFacade` handling, custom-comparer fallback, and a data-driven test asserting the set path and the scan path agree on 6 edge cases.
- Full `CoreTests` suite (net10.0): 1005 passed, 0 failed, 8 skipped.
- `Cql.Firely` builds with 0 warnings; a Release build confirms the public-API analyzer reports no new surface.

## FirelyTeam Checklist
- [x] **Update the title** of the PR to be succinct and less than 50 characters
- [ ] Mark the PR with the label **breaking change** when this PR introduces breaking changes (not applicable — no breaking changes)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_019LxWQZxuUUup1ZULebJnuK

### CODE REVIEW

File: docs/releases/vnext/1451-indexed-bundle-coding-cache.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+## Features
+
+- Retrieves with a code or value set filter over a `Bundle` no longer redo per-resource work on every
```
[the remark this answers is not in the document]
Comment (alexzautke) [reply docs/releases/vnext/1451-indexed-bundle-coding-cache.md:3]: Renamed in 1ca854b — to `1451-indexed-bundle-coding-cache.md` rather than `1450-…`: the README specifies the **PR** number, and #1450 is the issue this PR closes while the PR itself is #1451.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by baseTwo** [comment]
@copilot Reviewed this thoroughly — the caching/comparer/thread-safety logic all checks out correctly (traced the hash-set semantics against `StringCqlComparer`/`CqlComparer<T>.Equivalent`, confirmed `CompiledPropertyAccessor` bypassing `FhirModelPropertyInfo.GetValue`'s mapping path is safe, and verified cache-key scoping is collision-free across base-type retrieves). Two small non-blocking suggestions to apply before merge:

1. **Add a concurrency test.** The 25 new tests are all single-threaded. Since the memoization in `IndexedBundle`/`CompiledPropertyAccessor` is explicitly designed for concurrent access (a measure evaluation issues many retrieves over the same bundle), add a test that hits the same cache key (same type/property) from multiple threads concurrently and asserts no exceptions and consistent/correct results.
2. **Drop the trailing `#nullable disable`** in `Cql/Cql.Firely/CompiledPropertyAccessor.cs`. Pairing `#nullable enable`/`#nullable disable` isn't this project's convention (only one other, unrelated pre-existing file does it) — just remove the `#nullable disable` at the end of the file.

Please push both as a follow-up commit.

**Review by baseTwo — APPROVED** [review]
Both requested follow-ups landed cleanly: the two new concurrency tests (32 concurrent tasks against the same cache key, for both the primary-code-path and code-property paths) directly exercise the thread-safety claim, and the stray `#nullable disable` is gone. The rest of the logic (hash-set comparer semantics, cache-key scoping via `GetterIdentity`, `CompiledPropertyAccessor`'s bypass of `FhirModelPropertyInfo.GetValue`) was already verified correct in the prior pass and is unchanged here. Approving.
